internal/webhooks/cluster.go

```diff
@@ -395,6 +390,229 @@ func (webhook *Cluster) validateTopology(ctx context.Context, oldCluster, newClu
 }
 
+func (webhook *Cluster) validateTopologyVersion(ctx context.Context, fldPath *field.Path, fldValue interface{}, inVersion, oldVersion semver.Version, oldCluster *clusterv1.Cluster) field.ErrorList {
+	// Version could only be increased.
+	if inVersion.NE(semver.Version{}) && oldVersion.NE(semver.Version{}) && version.Compare(inVersion, oldVersion, version.WithBuildTags()) == -1 {
+		return field.ErrorList{field.Invalid(
+			fldPath,
+			fldValue,
+			fmt.Sprintf("version cannot be decreased from %q to %q", oldVersion, inVersion),
+		)}
+	}
+
+	// A +2 minor version upgrade is not allowed.
+	ceilVersion := semver.Version{
+		Major: oldVersion.Major,
+		Minor: oldVersion.Minor + 2,
+		Patch: 0,
+	}
+	if inVersion.GTE(ceilVersion) {
+		return field.ErrorList{field.Invalid(
+			fldPath,
+			fldValue,
+			fmt.Sprintf("version cannot be increased from %q to %q", oldVersion, inVersion),
+		),
+		}
+	}
+
+	// Only check the following cases if the minor version increases by at least 1.
+	ceilVersion = semver.Version{
+		Major: oldVersion.Major,
+		Minor: oldVersion.Minor + 1,
+		Patch: 0,
+	}
+
+	// Return early if its not a minor version upgrade.
+	if !inVersion.GTE(ceilVersion) {
+		return nil
+	}
+
+	allErrs := field.ErrorList{}
+	// minor version cannot be increased if KCP is upgrading or not yet on the current version
+	if err := validateTopologyControlPlaneVersion(ctx, fldPath, webhook.Client, oldCluster, oldVersion); err != nil {
+		allErrs = append(allErrs, field.Invalid(
+			fldPath,
+			fldValue,
+			fmt.Sprintf("blocking version update due to ControlPlane version check: %v", err),
+		))
+	}
+
+	// minor version cannot be increased if MachineDeployments are upgrading or not yet on the current version
+	if err := validateTopologyMachineDeploymentVersions(ctx, fldPath, webhook.Client, oldCluster, oldVersion); err != nil {
+		allErrs = append(allErrs, field.Invalid(
+			fldPath,
+			fldValue,
+			fmt.Sprintf("blocking version update due to MachineDeployment version check: %v", err),
+		))
+	}
+
+	// minor version cannot be increased if MachinePools are upgrading or not yet on the current version
+	if err := validateTopologyMachinePoolVersions(ctx, fldPath, webhook.Client, webhook.Tracker, oldCluster, oldVersion); err != nil {
+		allErrs = append(allErrs, field.Invalid(
+			fldPath,
+			fldValue,
+			fmt.Sprintf("blocking version update due to MachinePool version check: %v", err),
+		))
+	}
+
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+
+	return nil
+}
+
+func validateTopologyControlPlaneVersion(ctx context.Context, fldPath *field.Path, ctrlClient client.Reader, oldCluster *clusterv1.Cluster, oldVersion semver.Version) error {
+	cp, err := external.Get(ctx, ctrlClient, oldCluster.Spec.ControlPlaneRef, oldCluster.Namespace)
+	if err != nil {
+		return field.InternalError(fldPath, err)
+	}
+
+	cpVersionString, err := contract.ControlPlane().Version().Get(cp)
+	if err != nil {
+		return field.InternalError(fldPath, err)
```

Comment:
Should these here be regular errors so we don't nest field errors into field errors?
(same a few more times below)